contracts/Relay.sol

```diff
@@ -0,0 +1,431 @@
+                    /*
+                    NB: We store the height only every 4th header to save gas
+                    */
+                    blockHeight[_currentDigest] = _height;
+                }
+            }
+
+            /* NB: we do still need to make chain level checks tho */
+            require(_header.extractTarget() == _target, "Target changed unexpectedly");
+            require(_header.validateHeaderPrevHash(_previousDigest), "Headers not a consistent chain");
+
+            _previousDigest = _currentDigest;
+        }
+
+        emit Extension(
+            _anchor.hash256(),
+            _currentDigest);
+        return true;
+    }
+
+    /// @notice             Adds headers to storage after validating
+    /// @dev                We check integrity and consistency of the header chain
+    /// @param  _anchor     The header immediately preceeding the new chain
+    /// @param  _headers    A tightly-packed list of 80-byte Bitcoin headers
+    /// @return             True if successfully written, error otherwise
+    function addHeaders(bytes calldata _anchor, bytes calldata _headers) external returns (bool) {
+        return _addHeaders(_anchor, _headers, false);
+    }
+
+    /// @notice                 Adds headers to storage, performs additional validation of retarget
+    /// @dev                    Checks the retarget, the heights, and the linkage
+    /// @param  _oldPeriodStart The first header in the difficulty period being closed
+    /// @param  _oldPeriodEnd   The last header in the difficulty period being closed
+    /// @param  _headers        A tightly-packed list of 80-byte Bitcoin headers
+    /// @return                 True if successfully written, error otherwise
+    function addHeadersWithRetarget(
+        bytes calldata _oldPeriodStart,
+        bytes calldata _oldPeriodEnd,
+        bytes calldata _headers
+    ) external returns (bool) {
+        bytes memory _newPeriodStart = _headers.slice(0, 80);
+
+        /* NB: requires that both blocks are known */
+        uint256 _startHeight = _findHeight(_oldPeriodStart.hash256());
+        uint256 _endHeight = _findHeight(_oldPeriodEnd.hash256());
+
+        /* NB: retargets should happen at 2016 block intervals */
+        require(
+            _endHeight % 2016 == 2015,
+            "Must provide the last header of the closing difficulty period");
+        require(
+            _endHeight == _startHeight.add(2015),
+            "Must provide exactly 1 difficulty period");
+
+        /* NB: This comparison looks weird because header nBits encoding truncates targes */
+        uint256 _actualTarget = _newPeriodStart.extractTarget();
+        uint256 _expectedTarget = BTCUtils.retargetAlgorithm(
+            _oldPeriodStart.extractTarget(),
+            _oldPeriodStart.extractTimestamp(),
+            _oldPeriodEnd.extractTimestamp()
+        );
+        require(
+            (_actualTarget & _expectedTarget) == _actualTarget,
```

Comment:
What is this check for?

Reply:
Bitcoin difficulty is generated at full-bitlength, but only the 1.5 most-significant non-zero bytes are encoded into headers. So when doing the calculation, `_expectedTarget` uses all bits. We have to `AND` mask it with the blockheader-encoded target (`_actualTarget`) before making the equality comparison.

Basically we check if the observed encoded difficulty (pulled from the header) matches the expected encoded difficulty (calculated from the previous encoded difficulty and the timestamps).